### ISSUE TITLE
instantiation of formal parameters with only one symbol

### ISSUE THREAD

**Issue by mcodescu** [spechub/Hets#1846]
To support `SP[C]` instead of `SP[Class: C]`. Idea: introduce a new type of LIB_ITEM for unresolved IRIs. Then during static analysis of these new LIB_ITEMs attempt to download the file C (and solve the IRI as a download) and if that fails, instead of directly giving an error generate an actual parameter with a single symbol whose kind is determined by the kind of the only symbol of the formal parameter, if that makes sense (and solve the IRI as a symbol name).

**Comment by mcodescu** [comment]
if `C` is a spec declared in the same document, C is a `Spec_inst` and not a `Download_items`, and this will (and should) not change.

**Comment by tillmo** [comment]
agreed. Could you please implement this?

**Comment by mcodescu** [comment]
I have tried to do something, but I ran into some issues with generating the actual parameter specification with only one symbol that is the renaming of the unique symbol of the formal parameter. I have the symbol in the formal parameter as symbol in a certain logic `lid` and the name it should have in the actual parameter as an IRI. My idea was to build an EndoMap of raw symbols in `lid` using these, and then generate the fitting morphism using `induced_from_morphism` in `lid` (maybe you can see another way to do things).

In the CASL logic, I can then write List[Nat] where now Nat is not a specification, but the name of the sort Elem in the actual parameter, and induced_from_morphism generates the expected morphism. In OWL, however, when generating a raw symbol from the IRI, I get only an uri as a symbol (`AnUri`) instead of a symbol with kind (`ASymbol` with entity information). The case when a kinded symbol is mapped to an URI is not covered in `OWL2.Morphism.inducedFromMor` (and I'm under the impression that it is right not to do so), and thus an error is raised.

I am wondering whether we should not add in the class `Logic` a type of kinds of symbols, together with a method that takes a kind and an Id for the name (or even better an IRI) and generates a symbol of the given kind and with the given name.

Finally, I would expect that we should change the analyzed basic spec to a version where the unsolved IRIs have been solved, i.e. SP[C] becomes SP[Class:C]. In `Logic.Logic.StaticAnalysis` class there is a method `convertTheory` that takes a theory (a signature and a list of sentences) and gives a `basic_spec`, and I could use it to get a basic spec from the target signature of the fitting morphism together with the translation of the sentences of the formal parameter along the fitting morphism. But this is not implemented for any logic.

**Comment by tillmo** [comment]
The type of kinds of symbols sounds very reasonable (currently, kinds of symbols are represented as strings, which is less type-safe), and also the method generating a symbol.

Concerning the need of `convertTheory`, I do not fully understand your point. The current solution is that `basic_analysis` also returns a `basic_spec` where all kinds of things have been resolved. Doesn't this suffice?

**Comment by mcodescu** [comment]
Yes, but the question is, what exactly to put in the `basic_spec` that is returned as a result. 

Maybe a better solution than adding kinds is to have a method in the class `Logic` that takes a symbol and an IRI and returns a symbol that has as name the given IRI and inherits all other logic-dependent features a symbol may have (e.g. arity and result sorts, that are not given by kinds of symbols) from the given symbol.

**Comment by tillmo** [comment]
yes, this sounds good.

**Comment by mcodescu** [comment]
After introducing the method proposed in the last comment, the following spec is correctly analyzed by Hets:
```
logic OWL

spec S[Class: X] =
 Class: D SubClassOf: X
end

spec SC = S[C]
end
```
This only work because OWL is the only logic that does implement `convertTheory`. I need it to be able to get the basic spec `OntologyDocument {prefixDeclaration = fromList [], ontology = Ontology {name = , imports = [], ann = [], ontFrames = [Frame (ClassEntity (Expression C)) [AnnFrameBit [] (AnnotationFrameBit Declaration)]]}}` as an actual parameter in `S[C]` instead of initially `Spec_inst C [] Nothing nullRange`- as given by the parser, and then `Unsolved_IRI C` - as replaced before static analysis.

**Comment by tillmo** [comment]
good! I still do not fully understand the need of `convertTheory`. What do you excatly do with `Spec_inst C [] Nothing nullRange`?

**Comment by mcodescu** [comment]
If `C` is intended to be a symbol name in `S[C]` and not a specification name, `Spec_inst C [] Nothing nullRange` is not the correct `SPEC` for the argument of `S`. In such case, we want to generate the correct spec, which is of the form `Basic_spec G_basic_spec Range` with G_basic_spec containing the signature of `C`, and return it as return of analysis of `Unsolved_IRI C`, that substitutes `Spec_inst C [] Nothing nullRange` temporarily.

If your examples will only be in OWL, we can live with this.

**Comment by tillmo** [comment]
Yes, I understand all this. My question was: how do you come from `Spec_inst C [] Nothing nullRange` to a theory that you can use as input for `convertTheory`?

**Comment by mcodescu** [comment]
I've cleaned up the code, you can have a look at it here: https://github.com/spechub/Hets/pull/1849. There are things left to do, especially first trying to solve unknown IRIs as downloads in arguments of generic specs too.
